IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -450,24 +452,112 @@ - (BOOL)saveToken:(MSIDCredentialCacheItem *)credential
+    MSIDMacCredentialStorageItem *storageItem = [self syncStorageItem:key serializer:serializer context:context error:error];
+    [storageItem storeItem:credential forKey:key];
+    return [self saveStorageItem:storageItem key:key serializer:serializer context:context error:error];
+}
+
+// Read a single credential from the macOS keychain cache.
+// If multiple matches are found, return nil and set an error.
+- (MSIDCredentialCacheItem *)tokenWithKey:(MSIDCacheKey *)key
+                               serializer:(id<MSIDCacheItemSerializing>)serializer
+                                  context:(id<MSIDRequestContext>)context
+                                    error:(NSError **)error
+{
+    MSID_TRACE;
+    NSArray<MSIDCredentialCacheItem *> *items = [self tokensWithKey:key
+                                                         serializer:serializer
+                                                            context:context
+                                                              error:error];
+    
+    if (items.count > 1)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"The token cache store for this resource contains more than one token.");
+        [self createError:@"The token cache store for this resource contains more than one token."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorCacheMultipleUsers error:error context:context];
+        return nil;
+    }
+    
+    return items.firstObject;
+}
+
+// Read one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
+// If not found, return an empty list without setting an error.
+- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(MSIDCacheKey *)key
+                                           serializer:(id<MSIDCacheItemSerializing>)serializer
+                                              context:(id<MSIDRequestContext>)context
+                                                error:(NSError **)error
+{
+    NSArray *itemList = @[];
     
     /*
-     First step merge get latest from persistent cache and merge it with in-memory
-     Then write latest latest credential to in memory and write back to persistence
+     For refresh tokens, always merge with persistence to get the most recent refresh token as it is shared across apps from same publisher.
+     For AT/ID tokens, two apps sharing the same client id can write to the same entry to the keychain. In this case, it is possible that the first app is trying to read a credential which is not currently in its own memory but is written in persistence by the second app sharing the same client id. To find this credential, it is important to merge in memory cache with persistence.
      */
+    
+    MSIDMacCredentialStorageItem *storageItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
     MSIDMacCredentialStorageItem *savedStorageItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
```

Comment:
do we always need to read here? e.g. if item is not shared, we might as well have something in memory already. This code always reads even if we could retrieve it from memory.

Reply:
Good catch. Moved the check for savedStorageItem after the check for querying the in-momory cache. Applied same change for appMetadata and accountMetadata as they are also not shared.